Code Refactoring for getting start and stride from global ranks

shengfukevin opened the pull request:
Summary: Code Refactoring for getting start and stride from global ranks, this function can be used in different collective backend.

Differential Revision: D69555405




cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

kwen2501 commented:
Dumb q: how is `getGlobalRankStartAndStride` related to the title of the PR?

shengfukevin commented:
> Dumb q: how is `getGlobalRankStartAndStride` related to the title of the PR?

Sorry for the confusion. Since OSS version is exported from https://www.internalfb.com/diff/D69555405, and it only has the code related to OSS. The original DIFF is to capture HCCL in PyTorch ET. Since the code in getGlobalRankStartAndStride is reused by both HCCL and NCCL. I refactored it to a function. I can change the tile.

kwen2501 commented on torch/csrc/distributed/c10d/Backend.cpp:
```diff
@@ -15,4 +15,39 @@ void Backend::init() {
   C10_LOG_API_USAGE_ONCE(fmt::format("c10d.backend_{}", getBackendName()));
 }
 
+void getGlobalRankStartAndStride(
+    const std::vector<uint64_t>& globalRanksInGroup,
+    int& globalRankStart,
+    int& globalRankStride) {
+  if (globalRanksInGroup.empty()) {
+    globalRankStart = 0;
+  } else {
+    globalRankStart = static_cast<int>(globalRanksInGroup[0]);
+  }
+
+  if (globalRanksInGroup.empty()) {
+    globalRankStride = 1;
+  } else if (globalRanksInGroup.size() == 1) {
+    globalRankStride = 0;
+  } else {
+    bool ranksAreStrided = true;
+    auto startRank = globalRanksInGroup[0];
+    auto stride = globalRanksInGroup[1] - globalRanksInGroup[0];
+    for (std::vector<uint64_t>::size_type i = 0; i < globalRanksInGroup.size();
+         i++) {
+      if (globalRanksInGroup[i] != startRank + i * stride) {
+        ranksAreStrided = false;
+        break;
+      }
+    }
+
+    if (ranksAreStrided) {
+      globalRankStride =
+          static_cast<int>(globalRanksInGroup[1] - globalRanksInGroup[0]);
+    } else {
+      globalRankStride = -1;
+    }
+  }
+}
+
```
This function looks like a util function instead of a method of the `Backend` class. Shall we put it in `torch/csrc/distributed/c10d/Utils.hpp`?

shengfukevin replied:
fixed

kwen2501 commented on torch/csrc/distributed/c10d/Backend.hpp:
```diff
@@ -452,4 +452,9 @@ class TORCH_API Backend : public torch::CustomClassHolder {
   std::optional<at::Device> bound_device_id_;
 };
 
+void TORCH_API getGlobalRankStartAndStride(
+    const std::vector<uint64_t>& globalRanksInGroup,
+    int& globalRankStart,
+    int& globalRankStride);
+
```
nit: can you please document the arg fields and the purpose of this function? thanks.

shengfukevin replied:
fixed

kwen2501 reviewed (approved):
LGTM. Can you move the function and add documentation before landing? Thanks!

shengfukevin committed:
Capture HCCL in PyTorch Execution Trace (#147230)

Summary:

Following the same implementation of capturing NCCL in PyTorch execution trace, implement capturing HCCL.

Reviewed By: kwen2501, TiagoMAntunes

Differential Revision: D69555405
